plotbitrate.py

```diff
@@ -28,283 +29,596 @@
+) -> Generator[Frame, None, None]:
+    """
+    Creates an iterator from source_iterable and yields Frame objects.
+    """
+    for node in frame_elements(eTree.iterparse(source)):
+        # get data
+        time = try_get_frame_time_from_node(node)
+        size = node.get("pkt_size")
+        pict_type = node.get("pict_type")
+        # clear node to free parsed data
+        node.clear()
+
+        # construct and append frame
+        yield Frame(
+            time=time if time else 0,
+            size=int(size) if size else 0,
+            pict_type=pict_type if pict_type else "?"
+        )
+
+
+def frames_to_kbits(
+        frames: Iterable[Frame],
+        seconds_start: int,
+        seconds_end: int
+) -> Generator[Tuple[int, int], None, None]:
+    """
+    Creates a generator yielding every second between seconds_start
+    and seconds_end (including both) and its summed size in kbit.
+
+    The frames iterable must be sorted by frame time.
+    """
+    frames_iter = iter(frames)
+    last_frame_second = 0
+    last_frame_size = 0
+
+    # loop over every second
+    for second in range(seconds_start, seconds_end + 1):
+
+        # restore size of a saved frame from last iteration
+        # if it's for the current second
+        if last_frame_second == second:
+            size = last_frame_size
+        else:
+            size = 0
+
+        # advance iterator only if the saved frame data
+        # is not for a future second
+        if last_frame_second <= second:
+            # advances the iterator until it's at a frame
+            # belonging to a future second
+            for frame in frames_iter:
+                frame_second = math.floor(frame.time)
+                if frame_second < second:
+                    continue
+                elif frame_second == second:
+                    # frame is current second, so sum up
+                    size += frame.size
+                else:
+                    # current frame is not in current second
+                    # store its size and second and break iteration
+                    last_frame_second = frame_second
+                    last_frame_size = frame.size
+                    break
+
+        yield second, int(size * 8 / 1000)
+
+
+def downscale_bitrate(
+        bitrates: Dict[int, int],
+        factor: int
+) -> Generator[Tuple[int, int], None, None]:
+    """
+    Groups bitrates together and takes the highest bitrate as the value.
+
+    Args:
+        bitrates: dict containing seconds with bitrates
+        factor: which seconds to keep (1 is every, 2 is every other and so on)
+
+    Example:
+
+        given the parameters:
+
+        bitrates = {
+            0: 3400,
+            1: 5290
+            2: 4999
+            3: 7500
+            4: 0
+            5: 7800
+            6: 3000
+        }
+        factor = 3
+
+        this function will return an iterator giving:
+        (0, 5290)
+        (3, 7800)
+        (6, 3000)
```

Comment:
That's a good question. I don't know since it's closed source. But doing some comparisons and trying to understand what it does I came up with this behavior.
Bitrate viewer always shows peak values. If I average them the spikes get hidden and I wanted to avoid that because the peaks are what I want to see the most.
I don't really know how useful this is for others. It's something I wanted to have for myself but I'm sure there are better ways. I could make an additional command line switch but that feels convoluted.
What's your opinion?

Reply:
I see value in both approaches, so my opinion is either option would be fine. Since you have a personal use case for the maximum let's just keep it at that.